blog/_posts/2017-02-25-scalafix-v0.3.md

```diff
@@ -0,0 +1,154 @@
+---
+layout: blog
+post-type: blog
+by: Ólafur Páll Geirsson
+title: "Refactor with scalafix v0.3"
+---
+
+I am happy to announce the release of [scalafix v0.3][scalafix].
+This release uses the new scala.meta semantic API to provide a re-designed `Rewrite` and `Patch` API for refactorings.
+Let me explain what that means word by word.
+
+> Note. This is the second post on scalafix and scala.meta. You might be
+> interested in reading the first post
+> [here](http://scala-lang.org/blog/2016/10/24/scalafix.html).
+
+## Scala.meta semantic API
+
+Scala.meta recently announced the [first release of its semantic API][meta-1.6].
+This release is the product of close collaboration for the past several months between
+[@xeno_by][] at Twitter and [myself][@olafurpg] at the Scala Center.
+The objective of the semantic API is to provide operations to query information from the compiler.
+
+The first version of the scala.meta semantic API makes it possible to query for the resolved "symbol" of a name that appears in a Scala source file.
+A name is a reference to some definition, for example `println` or `scala.Predef.println`.
+A symbol is a unique identifier of a single definition.
+For example, `println` from the standard library has the symbol `_root_.scala.Predef.println(Ljava/lang/Object;)V.`.
+The compiler is responsible for resolving names to symbols.
+
+Scalahost is a compiler plugin in the scala.meta project that extracts symbols from the compiler and maps them to scala.meta syntax trees.
+Scalahost emits the extracted symbols into a "semantic database".
+The semantic database can be persisted to files on disk and loaded for later analysis.
+Semantic databases from different compilation units, potentially produced by different
+versions of the Scala compiler, can be merged.
+This opens possibilities for large-scale code analysis.
+
+The introduction of the scala.meta semantic API is a game changer for scalafix.
+The ability to resolve names to symbols opens possibilities for many scalafix rewrites.
+
+## Rewrite: meta.Tree => Seq[Patch]
+
+A scalafix `Patch` is a small operation that can produce a diff on a Scala source file.
+A patch can either be a "token patch" or a "tree patch".
+
+Token patches are low-level but give full control over how every detail in a source file is handled, for example formatting and comments.
+Example token patches are `Remove(token)` and `AddLeft(token, toAdd: String)`, which removes or prepends a string to `token`, respectively.
+
+Tree patches are high-level and allow rewrite author to declaratively explain what operation to perform.
+An example tree patch is `AddGlobalImport(importer)`, which adds a new import to the top of a file if it does not exist.
+Observe that `AddGlobalImport` does not worry about token-level details such as whether the user groups imports by prefix (e.g., `import a.{b, c}`).
```

Comment:
I clarified this part. I didn't mention import rename since that is not really handled at the moment.